### ISSUE TITLE
pkg-config install fail. `configure: error: C compiler cannot create executables`

### ISSUE THREAD

**Issue by macmanes** [Linuxbrew/brew#200]
# Please follow the general troubleshooting steps first:

- [x] Ran `brew update` and retried your prior step?
- [x] Ran `brew doctor`, fixed as many issues as possible and retried your prior step?

### Bug reports:

https://gist.github.com/anonymous/ccf33571265d355163362e2402c9b09d

I am trying to install several packages, with failures related to `pkg-config`, and am getting errors (see gist), which I believe are fundamentally related to `configure: error: C compiler cannot create executables`

I've looked at a few other bug report of this nature, and they seem to be a bit, ahem, complicated.

**Comment by sjackman** [comment]
The relevant error is
```
/usr/bin/ld: /mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/lib/../lib64/crti.o: unrecognized relocation (0x2a) in section `.init'
/usr/bin/ld: final link failed: Bad value
```
https://gist.github.com/anonymous/ccf33571265d355163362e2402c9b09d#file-config-log-L98

**Comment by sjackman** [comment]
```
OS: Scientific Linux release 7.2 (Nitrogen)
```

Try `brew install binutils`

**Comment by macmanes** [comment]
seems like binutils already installed installed:

```
brew install binutils
Updating Homebrew...
==> Homebrew has enabled anonymous aggregate user behaviour analytics.
Read the analytics documentation (and how to opt-out) here:
  https://git.io/brew-analytics

==> Auto-updated Homebrew!
Updated 1 tap (homebrew/core).
No changes to formulae.

Warning: binutils-2.27 already installed


[macmanes@premise linuxbrew]$ brew install pkg-config
==> Downloading https://pkgconfig.freedesktop.org/releases/pkg-config-0.29.1.tar.gz
Already downloaded: /mnt/lustre/macmaneslab/macmanes/.cache/Homebrew/pkg-config-0.29.1.tar.gz
==> ./configure --prefix=/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Cellar/pkg-config/0.29.1_2 --disable-host-tool --with-internal-glib --with-pc-path=/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/lib/pkgconfig:/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/share/pkgconfig:/usr/local/lib/pkg
Last 15 lines from /mnt/lustre/macmaneslab/macmanes/.cache/Homebrew/Logs/pkg-config/01.configure:
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /usr/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking build system type... x86_64-unknown-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking how to print strings... printf
checking for style of include used by make... GNU
checking for gcc... /mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/bin/gcc-5
checking whether the C compiler works... no
configure: error: in `/tmp/pkg-config-20161207-98803-ibkli8/pkg-config-0.29.1':
configure: error: C compiler cannot create executables
See `config.log' for more details

READ THIS: https://github.com/Linuxbrew/brew/blob/master/docs/Troubleshooting.md#troubleshooting

```

**Comment by sjackman** [comment]
You'll need to troubleshoot why `pkg-config` is picking up `/usr/bin/ld` rather than `$(brew --prefix)/bin/ld`. Try
```
/usr/bin/ld --version
brew link binutils
ld --version
which -a ld
brew sh <<<'which -a ld'
brew install -dv pkg-config
```
select `shell` then run `which -a ld`

**Comment by sjackman** [comment]
You want to see in `config.log`:
```
configure:4406: checking for ld used by /home/sjackman/.linuxbrew/bin/gcc-5
configure:4473: result: /home/sjackman/.linuxbrew/bin/ld
```
but your run of `configure` is not getting that far.

**Comment by sjackman** [comment]
Please report the output of
```sh
$(brew --prefix gcc)/libexec/gcc/x86_64-unknown-linux-gnu/5.3.0/collect2 --version
```

**Comment by macmanes** [comment]
```
$(brew --prefix gcc)/libexec/gcc/x86_64-unknown-linux-gnu/5.3.0/collect2 --version

collect2 version 5.3.0
/usr/bin/ld --version
GNU ld version 2.23.52.0.1-55.el7 20130226
Copyright 2013 Free Software Foundation, Inc.
This program is free software; you may redistribute it under the terms of
the GNU General Public License version 3 or (at your option) a later version.
This program has absolutely no warranty.

```

**Comment by macmanes** [comment]
not sure if I'm doing these steps exactly right, so let me know. 

```
which ld
~/linuxbrew/.linuxbrew/bin/ld
```

```
[macmanes@premise linuxbrew]$ /usr/bin/ld --version
GNU ld version 2.23.52.0.1-55.el7 20130226
```

```
brew link binutils
Warning: Already linked: /mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Cellar/binutils/2.27
To relink: brew unlink binutils && brew link binutils
[macmanes@premise linuxbrew]$ ld --version
GNU ld (GNU Binutils) 2.27
```

```
which -a ld
~/linuxbrew/.linuxbrew/bin/ld
~/linuxbrew/.linuxbrew/bin/ld
/usr/bin/ld
```

```
[macmanes@premise linuxbrew]$ brew sh <<<'which -a ld'
Your shell has been configured to use Homebrew's build environment:
this should help you build stuff. Notably though, the system versions of
gem and pip will ignore our configuration and insist on using the
environment they were built under (mostly). Sadly, scons will also
ignore our configuration.
When done, type `exit'.
/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/bin/ld
/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/bin/ld
/usr/bin/ld


[macmanes@premise linuxbrew]$ brew install -dv pkg-config
/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Library/Taps/homebrew/homebrew-core/Formula/pkg-config.rb
/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Library/Homebrew/build.rb (Formulary::FromPathLoader): loading /mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Library/Taps/homebrew/homebrew-core/Formula/pkg-config.rb
==> Downloading https://pkgconfig.freedesktop.org/releases/pkg-config-0.29.1.tar.gz
Already downloaded: /mnt/lustre/macmaneslab/macmanes/.cache/Homebrew/pkg-config-0.29.1.tar.gz
==> Verifying pkg-config-0.29.1.tar.gz checksum
tar xzf /mnt/lustre/macmaneslab/macmanes/.cache/Homebrew/pkg-config-0.29.1.tar.gz
==> ./configure --disable-debug --prefix=/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Cellar/pkg-config/0.29.1_2 --disable-host-tool --with-internal-glib --with-pc-path=/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/lib/pkgconfig:/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/share/pkgconfig:/usr/local/lib/pkgconfig:/usr/lib/pkgconfig:/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Library/Homebrew/os/mac/pkgconfig/0
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /usr/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking build system type... x86_64-unknown-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking how to print strings... printf
checking for style of include used by make... GNU
checking for gcc... /mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/bin/gcc-5
checking whether the C compiler works... no
configure: error: in `/tmp/pkg-config-20161207-112122-q3b44z/pkg-config-0.29.1':
configure: error: C compiler cannot create executables
See `config.log' for more details
/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Library/Taps/homebrew/homebrew-core/Formula/glibc.rb
/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Library/Taps/homebrew/homebrew-core/Formula/gcc.rb
/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Library/Homebrew/build.rb (Formulary::TapLoader): loading /mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Library/Taps/linuxbrew/homebrew-xorg/xorg.rb
/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Library/Homebrew/debrew.rb:11:in `raise'
FormulaUnavailableError: No available formula with the name "xorg"
1. raise
2. ignore
3. backtrace
4. irb
5. shell
Choose an action: 5
When you exit this shell, you will return to the menu.

bash-4.2$ which -a ld
/usr/bin/ld
```

**Comment by macmanes** [comment]
PS, ths doesn't have anything to do with the symlinking of the locks folder to `/tmp`, right?

**Comment by sjackman** [comment]
I'm pretty sure it's unrelated.

**Comment by sjackman** [comment]
> `FormulaUnavailableError: No available formula with the name "xorg"`

Run `brew tap linuxbrew/xorg`

**Comment by macmanes** [comment]
new error after xorg installed:

```
 brew install -dv pkg-config
/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Library/Taps/homebrew/homebrew-core/Formula/pkg-config.rb
/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Library/Homebrew/build.rb (Formulary::FromPathLoader): loading /mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Library/Taps/homebrew/homebrew-core/Formula/pkg-config.rb
==> Downloading https://pkgconfig.freedesktop.org/releases/pkg-config-0.29.1.tar.gz
Already downloaded: /mnt/lustre/macmaneslab/macmanes/.cache/Homebrew/pkg-config-0.29.1.tar.gz
==> Verifying pkg-config-0.29.1.tar.gz checksum
tar xzf /mnt/lustre/macmaneslab/macmanes/.cache/Homebrew/pkg-config-0.29.1.tar.gz
==> ./configure --disable-debug --prefix=/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Cellar/pkg-config/0.29.1_2 --disable-host-tool --with-internal-glib --with-pc-path=/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/lib/pkgconfig:/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/share/pkgconfig:/usr/local/lib/pkgconfig:/usr/lib/pkgconfig:/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Library/Homebrew/os/mac/pkgconfig/0
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /usr/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking build system type... x86_64-unknown-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking how to print strings... printf
checking for style of include used by make... GNU
checking for gcc... /mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/bin/gcc-5
checking whether the C compiler works... no
configure: error: in `/tmp/pkg-config-20161207-114949-s93d5k/pkg-config-0.29.1':
configure: error: C compiler cannot create executables
See `config.log' for more details
/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Library/Taps/homebrew/homebrew-core/Formula/glibc.rb
/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Library/Taps/homebrew/homebrew-core/Formula/gcc.rb
/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Library/Homebrew/build.rb (Formulary::TapLoader): loading /mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Library/Taps/linuxbrew/homebrew-xorg/xorg.rb
/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Library/Homebrew/debrew.rb:11:in `raise'

BuildError: Failed executing: ./configure --disable-debug --prefix=/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Cellar/pkg-config/0.29.1_2 --disable-host-tool --with-internal-glib --with-pc-path=/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/lib/pkgconfig:/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/share/pkgconfig:/usr/local/lib/pkgconfig:/usr/lib/pkgconfig:/mnt/lustre/macmaneslab/macmanes/linuxbrew/.linuxbrew/Library/Homebrew/os/mac/pkgconfig/0

```

**Comment by sjackman** [comment]
```
$(brew --prefix gcc)/libexec/gcc/x86_64-unknown-linux-gnu/5.3.0/collect2 --version

collect2 version 5.3.0
/usr/bin/ld --version
```

That's not good, but I'm not sure how to fix it. We want it to report Linuxbrew's `ld`. Here's the relevant documentation
https://gcc.gnu.org/onlinedocs/gccint/Collect2.html

Please report `which -a real-ld`

**Comment by sjackman** [comment]
```
5. shell
Choose an action: 5
When you exit this shell, you will return to the menu.

bash-4.2$ which -a ld
/usr/bin/ld
```
That's the smoking gun! That shouldn't happen. Please report
```sh
brew install -dv pkg-config
```
Select `shell` and report `echo $PATH`

Any chance either your `~/.bashrc` or `~/.bash_profile` puts `/usr/bin` in your `PATH` before `$(brew --prefix)/bin` ?

**Comment by emreerhan** [comment]
I was having a similar problem:

```
$ brew install fontconfig openblas libpthread-stubs numpy scipy
==> Installing dependencies for fontconfig: pkg-config, libpng, freetype
==> Installing fontconfig dependency: pkg-config
==> Downloading https://pkgconfig.freedesktop.org/releases/pkg-config-0.29.2.tar.gz
Already downloaded: /home/eerhan/.cache/Homebrew/pkg-config-0.29.2.tar.gz
==> ./configure --prefix=/home/eerhan/.linuxbrew/Cellar/pkg-config/0.29.2 --disable-host-tool --with-internal-glib --
Last 15 lines from /home/eerhan/.cache/Homebrew/Logs/pkg-config/01.configure:
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking build system type... x86_64-unknown-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking how to print strings... printf
checking for style of include used by make... GNU
checking for gcc... gcc-5
checking whether the C compiler works... no
configure: error: in '/tmp/pkg-config-20170425-148819-q05kra/pkg-config-0.29.2':
configure: error: C compiler cannot create executables
See 'config.log' for more details
```

But it was fixed when I reinstalled binutils with:
```
$brew unlink binutils
$brew install binutils
$brew install -dv fontconfig
```

Everything seems to work now. I just thought it was worth documenting.

**Comment by sjackman** [comment]
I believe that this issue is resolved by both #272 and #355.
#355 probably means that #272 could be reverted.
